core/templates/pages/exploration-player-page/services/conversation-skin.service.ts

```diff
@@ -0,0 +1,155 @@
+})
+export class ConversationSkinService {
+  // The minimum width, in pixels, needed to be able to show two cards
+  // side-by-side.
+  TIME_PADDING_MSEC = 250;
+  TIME_FADEIN_MSEC = 100;
+  TIME_NUM_CARDS_CHANGE_MSEC = 500;
+
+  isAnimatingToTwoCards!: boolean;
+  isAnimatingToOneCard!: boolean;
+
+  constructor(
+    private contentTranslationLanguageService: ContentTranslationLanguageService,
+    private contentTranslationManagerService: ContentTranslationManagerService,
+    private explorationPlayerStateService: ExplorationPlayerStateService,
+    private playerPositionService: PlayerPositionService,
+    private playerTranscriptService: PlayerTranscriptService,
+    private windowDimensionsService: WindowDimensionsService
+  ) {}
+
+  handleNewCardAddition(newCard: StateCard): void {
+    this.playerTranscriptService.addNewCard(newCard);
+    const explorationLanguageCode =
+      this.explorationPlayerStateService.getLanguageCode();
+    const selectedLanguageCode =
+      this.contentTranslationLanguageService.getCurrentContentLanguageCode();
+    if (explorationLanguageCode !== selectedLanguageCode) {
+      this.contentTranslationManagerService.displayTranslations(
+        selectedLanguageCode
+      );
+    }
+
+    let totalNumCards = this.playerTranscriptService.getNumCards();
+
+    let previousSupplementalCardIsNonempty =
+      totalNumCards > 1 &&
+      this.isSupplementalCardNonempty(
+        this.playerTranscriptService.getCard(totalNumCards - 2)
+      );
+
+    let nextSupplementalCardIsNonempty = this.isSupplementalCardNonempty(
+      this.playerTranscriptService.getLastCard()
+    );
+
+    if (
+      totalNumCards > 1 &&
+      this.canWindowShowTwoCards() &&
+      !previousSupplementalCardIsNonempty &&
+      nextSupplementalCardIsNonempty
+    ) {
+      this.playerPositionService.setDisplayedCardIndex(totalNumCards - 1);
+      this.animateToTwoCards(function () {});
+    } else if (
+      totalNumCards > 1 &&
+      this.canWindowShowTwoCards() &&
+      previousSupplementalCardIsNonempty &&
+      !nextSupplementalCardIsNonempty
+    ) {
+      this.animateToOneCard(() => {
+        this.playerPositionService.setDisplayedCardIndex(totalNumCards - 1);
+      });
+    } else {
+      this.playerPositionService.setDisplayedCardIndex(totalNumCards - 1);
+    }
+    this.playerPositionService.changeCurrentQuestion(
+      this.playerPositionService.getDisplayedCardIndex()
+    );
+  }
+
+  // Returns whether the screen is wide enough to fit two
+  // cards (e.g., the tutor and supplemental cards) side-by-side.
+  canWindowShowTwoCards(): boolean {
+    return (
+      this.windowDimensionsService.getWidth() >
+      ExplorationPlayerConstants.TWO_CARD_THRESHOLD_PX
+    );
+  }
+
+  isPlayerAnimatingToOneCard(): boolean {
+    return this.isAnimatingToOneCard;
+  }
+
+  isPlayerAnimatingToTwoCards(): boolean {
+    return this.isAnimatingToTwoCards;
+  }
+
+  animateToTwoCards(doneCallback: () => void): void {
```

Comment:
Hmm, is this callback here only for the tests? That's not a good sign. If I'm correct, please remove it and try to adjust the tests

Reply:
It's from the current code and doesn't look like it's for tests, we pass the appropriate `playerPositionService.setDisplayedCardIndex()` function as the callback and have it executed after the set timeout when the animation's done.